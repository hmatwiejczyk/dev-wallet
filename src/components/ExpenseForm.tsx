import * as React from 'react';
import * as moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';
import 'react-dates/lib/css/_datepicker.css';

class ExpenseForm extends React.Component<
  { onSubmit?; expense? },
  {
    description: string;
    amount: string;
    note: string;
    createdAt: any;
    calendarFocused: boolean;
    errors: { descriptionError; amountError };
  }
> {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? props.expense.amount.toString() : '',
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      errors: {
        descriptionError: '',
        amountError: ''
      }
    };
  }
  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onNoteChange = e => {
    e.persist();
    this.setState(() => ({ note: e.target.value }));
  };
  onDateChange = createdAt => {
    if (createdAt) this.setState(() => ({ createdAt }));
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      let descriptionError = '';
      let amountError = '';
      if (!this.state.description) {
        descriptionError = 'This field is required';
      }
      if (!this.state.amount) {
        amountError = 'This field is required';
      }
      this.setState(() => ({ errors: { descriptionError, amountError } }));
    } else {
      this.setState(() => ({
        errors: { amountError: '', descriptionError: '' }
      }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount),
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };
  render() {
    return (
      <div>
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          {this.state.errors.descriptionError && (
            <small>{this.state.errors.descriptionError}</small>
          )}
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          {this.state.errors.amountError && (
            <small>{this.state.errors.amountError}</small>
          )}
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            displayFormat={'DD/MM/YY'}
            isOutsideRange={() => false}
          />
          <textarea
            cols={30}
            rows={10}
            placeholder="Add more for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>

          <button type="submit">Add Expense</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
