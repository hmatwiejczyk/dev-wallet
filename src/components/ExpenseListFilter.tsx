import * as React from 'react';
import { connect } from 'react-redux';
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from '../actions/filters';
import { DateRangePicker } from 'react-dates';
interface IFilterProps {
  filters;
  dispatch;
}
interface IFilterState {
  startDate?;
  endDate?;
  focusedInput?;
  calendarFocused?;
}
class ExpenseListFilter extends React.Component<IFilterProps, IFilterState> {
  state = {
    calendarFocused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  };
  onFocusChange = calendarFocused => {
    this.setState(() => ({ calendarFocused }));
  };
  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={e => {
            this.props.dispatch(setTextFilter(e.target.value));
          }}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={e => {
            if (e.target.value === 'date') {
              this.props.dispatch(sortByDate());
            } else if (e.target.value === 'amount') {
              this.props.dispatch(sortByAmount());
            }
          }}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          startDateId="your_unique_start_date_id"
          endDate={this.props.filters.endDate}
          endDateId="your_unique_end_date_id"
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          displayFormat={'DD/MM/YY'}
          isOutsideRange={() => false}
          showClearDates
        />
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    filters: state.filters
  };
};

export default connect(mapStateToProps)(ExpenseListFilter);
