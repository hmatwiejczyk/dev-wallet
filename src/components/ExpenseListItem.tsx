import * as React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';

const ExpenseListItem = ({ dispatch, id, description, amount, createdAt }) => (
  <div>
    <h4>Description: {description}</h4>
    <small>
      Amount: {amount} | Created: {createdAt}
      <br />
      <button onClick={() => dispatch(removeExpense({ id }))}>Remove</button>
    </small>
  </div>
);

export default connect()(ExpenseListItem);
