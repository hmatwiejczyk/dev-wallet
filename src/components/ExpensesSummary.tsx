import * as React from 'react';
import { connect } from 'react-redux';
import getTotalExpenses from '../selectors/expenses-total';
import getVisibleExpenses from '../selectors/expenses';

export const ExpensesSummary = props => {
  const expenseWord = props.expensesCount === 1 ? 'expense' : 'expenses';
  return (
    <div>
      Viewing {props.expensesCount} {expenseWord} totalling {props.totalExpenses}
    </div>
  );
};

const mapStateToProps = state => {
  return {
    totalExpenses: getTotalExpenses(
      getVisibleExpenses(state.expenses, state.filters)
    ),
    expensesCount: state.expenses.length
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
