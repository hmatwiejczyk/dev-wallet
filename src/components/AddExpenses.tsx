import * as React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';

export const AddExpensePage = props => (
  <div>
    <h2>Add Expenses</h2>
    <ExpenseForm
      onSubmit={expense => {
        // props.dispatch(addExpense(expense));
        props.onSubmit(expense);
        props.history.push('/');
      }}
    />
  </div>
);

const mapDispatchToProps = dispatch => ({
  onSubmit: expense => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
