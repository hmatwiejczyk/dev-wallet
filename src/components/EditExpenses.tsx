import * as React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense, removeExpense } from '../actions/expenses';

const EditExpensePage = props => {
  return (
    <div>
      <ExpenseForm
        expense={props.expense}
        onSubmit={expense => {
          props.dispatch(addExpense(expense));
          props.history.push('/');
        }}
      />
      <button
        onClick={() => {
          props.dispatch(removeExpense(props.expense));
          props.history.push('/');
        }}
      >
        Remove
      </button>
    </div>
  );
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      expense => expense.id === props.match.params.id
    )
  };
};

export default connect(mapStateToProps)(EditExpensePage);
