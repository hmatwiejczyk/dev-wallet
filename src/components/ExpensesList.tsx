import * as React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../selectors/expenses';

const ExpensesList = props => (
  <div>
    <h2>Expense List</h2>
    <div>
      {props.expenses.map((e, i) => (
        <ExpenseListItem key={e.id} {...e} />
      ))}
    </div>
  </div>
);

const mapStateToProps = state => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpensesList);
