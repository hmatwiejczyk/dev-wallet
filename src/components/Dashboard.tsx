import * as React from 'react';
import ExpensesList from './ExpensesList';
import ExpenseListFilter from './ExpenseListFilter';

export const Dashboard = () => (
  <div>
    <ExpenseListFilter />
    <ExpensesList />
  </div>
);
