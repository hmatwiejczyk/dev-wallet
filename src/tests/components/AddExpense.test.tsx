import * as React from 'react';
import { shallow } from "enzyme";
import { AddExpensePage } from "../../components/AddExpenses";

let onSubmit, history, wrapper;

beforeEach(() => {
   onSubmit = jest.fn();
   history = { push: jest.fn() }
   wrapper = shallow(<AddExpensePage onSubmit={onSubmit} history={history}/>)
})

test('should render AddExpense correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
  const expense = {
        id: 1,
        description: 'test 1',
        note: 'test 1',
        amount: 1000,
        createdAt: 333
      };
  wrapper.find('ExpenseForm').prop('onSubmit')(expense as any);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(onSubmit).toHaveBeenLastCalledWith(expense);
})

