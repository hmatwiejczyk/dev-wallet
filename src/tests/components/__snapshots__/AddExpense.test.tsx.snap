// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExpense correctly 1`] = `
<div>
  <h2>
    Add Expenses
  </h2>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
